---
title: "METALGEN_SQMr"
author: "Adrian Lopez Garcia"
date: '`r format(Sys.time(), "%d/%m/%Y")`'
output: html_document
---

```{r setup, include=FALSE}
dic.code <- "C:/Users/INIA/5_R_Scripts"

knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)

packs <- c("easyCODA", "ggordiplots", "gridExtra", "kableExtra", "knitr", "limma", "microbiome", "phyloseq", "plotly", 
           "propr", "sunburstR", "tidyverse", "vegan", "zCompositions")
lapply(packs, require, character.only = TRUE)
library(microDA)

savefiles <- FALSE

count_distrib <- function(data){
  counts_dist <- NULL
  # FUNDAMENTAL COUNTS VALUES: 0, 1, 2, 3-10, 10-100, 100-1000, >1000
  for(i in seq(ncol(data))){
    zero <- sum(data[,i] == 0)
    one <- sum(data[,i] == 1)
    two <- sum(data[,i] == 2)
    r10 <- sum(data[,i] >= 3 & data[,i] < 10)
    r100 <- sum(data[,i] >= 10 & data[,i] < 100)
    r1000 <- sum(data[,i] >= 100 & data[,i] < 1000)
    more <- sum(data[,i] >= 1000)
    row <- cbind(zero, one, two, r10, r100, r1000, more)
    counts_dist <- rbind(counts_dist, row)
    rownames(counts_dist)[i] <- colnames(data)[i]
    rm(zero, one, two, r10, r100, r1000, more, row)
  }
  counts_dist <- as.data.frame(counts_dist) %>% rownames_to_column() %>% dplyr::arrange(zero, one, two)
  colnames(counts_dist)[1] <- "samples"
  # Data frame containing order index per sample (based on number of zeros):
  counts_dist_sort <- as.data.frame(cbind("order" = as.numeric(row.names(counts_dist)), "samples" = counts_dist$samples), stringsAsFactors = F)
  # Long data format and merge with sample rder index:
  counts_dist_ggdat_pre <- reshape2::melt(counts_dist, id.vars = "samples")
  counts_dist_ggdat <- merge(counts_dist_ggdat_pre, counts_dist_sort, by = "samples")
  counts_dist_ggdat$order <- as.numeric(counts_dist_ggdat$order)
  # GGPLOT:
  if (nrow(data) < 500) {syseq <- 200}
  else syseq <- 500
  ggcounts <- ggplot(counts_dist_ggdat) + 
    aes(x = reorder(samples, -order), y = value, fill = variable) + 
    geom_bar(stat = "identity", width = 1, position = position_stack(reverse = T)) +
    xlab("Samples") + ylab("Number of features") +
    theme(axis.text.x = element_blank(), axis.title = element_text(size=14, face="bold")) +
    scale_y_continuous(breaks = seq(0, nrow(data), syseq), expand = expansion(mult = c(0, 0.03))) +
    scale_fill_discrete(name = "Counts per feature", labels = c("Zero", "Singletons", "Doubletons", "3 to 10", "10 to 100", "100 to 1000", "1000 +"))
  return(ggcounts)
}

```

```{r init_variables, include=FALSE, paged.print=FALSE, eval=TRUE}
## WORKING DIRECTORY & RSAVES
WD <- "D:/Adrian/METALGEN/MinION/"
RSAV <- paste0(WD, "Rsaves/Rdata/")

ifelse(!dir.exists(file.path(RSAV)), dir.create(file.path(RSAV)), FALSE)

## INPUT VARIABLES
# metadataset: Logical control variable. If TRUE metadata will be required. If FALSE metadata can be NULL.
# taxdenom: Names assigned for taxa ranges.
# taxlvl: Level for classification limit (by default: 7 = species)
# taxpattern: Regex pattern present in taxonomic ref. database. It will be removed during phyloseq taxa fill
#   Possible values: "\\[\\w+\\] ", "\\w_\\d__"
metadataset <- TRUE
taxdenom <- c("superkingdom", "phylum", "class", "order", "family", "genus", "species")
taxlvl <- 6
taxrng <- seq(taxlvl)
#taxDAlvls <- taxdenom[-c(1,7)]
#taxpattern <- "\\w_"

## ANALYSIS PHENOT VARIABLES:
phenots <- c("ch4.spkpm_adjusted", "NLACTA", "DIASLE", "CH4_Rank")

## OUTLIERS:
#outliers <- c("SH_3", "SH_1", "AMET_016")
outliers <- c("SH_3", "AMET_016")

## PLOTTING VARIABLES
prevalence.tax <- "Phylum"

## DA VARIABLES
# Not limma:
alpha = 0.05
# Limma variables:
fc.cut = 0.5
qval = 0.05

```

```{r dataset_loading, include=FALSE, paged.print=FALSE, eval=TRUE}
## DATA LOADING
# Counts table (tipically output from data_filter.R script), filtered by samples, taxonomy and prevalence.
data.tx <- read.csv(paste0(WD,"SQMReads/SQMr_filtered_genera.csv"), sep = ",", dec = ".", stringsAsFactors = F, row.names = 1)
data.kg <- read.delim(paste0(WD,"SQMReads/SQMr_filtered_kegg.csv"), stringsAsFactors = F, row.names = 1)

## METADATA
# Optional metadata .csv. It must contain sample IDs plus optional sample groups.
if (metadataset == TRUE){
  ids_pre <- read.csv(paste0(WD, "metadata.csv"), sep = ";", dec = ",", stringsAsFactors = F, row.names = 1)
  if(all.equal(colnames(data.tx)[-c(1:taxlvl)], colnames(data.kg)[-tail(1:ncol(data.kg),2)])){
    ids <- ids_pre[rownames(ids_pre) %in% colnames(data.tx),]
  } else {stop()}
} else if (metadataset == FALSE){
  ids <- data.frame("ID"=colnames(dat[,-taxrng]), "Group"=rep("No group", NCOL(dat[,-taxrng])))
  row.names(ids) <- colnames(dat[,-taxrng])
}

# Metadata pruning & editing:
ids_out <- ids
ids <- ids[!(rownames(ids) %in% outliers),]

ids$CH4_Rank <- cut(ids$ch4.spkpm_adjusted, breaks = c(quantile(ids$ch4.spkpm_adjusted, prob = c(0, 0.25, 0.50, 0.75, 1))),
                    labels=c("LOW","L_MID","H_MID","HIGH"), include.lowest = TRUE)
ids$NLACTA <- as.factor(ids$NLACTA)
ids$DIASLE <- cut(ids$DIASLE, breaks = c(0, 70, 150, 1000), labels = c("<70", "70-150", ">150"), right = FALSE)

# Save metadata:
if(isTRUE(savefiles)){
  save(ids_pre, ids_out, ids, file = paste0(RSAV, "00_METADATA.rda"))
}

```

# TAXONOMY DATASET PROCESSING:

```{r phyloseq, include=FALSE}
## DATASET PREPARATION
# Separation of data table into taxonomy and counts tables:
data.tx_reads <- data.tx[,-which(colnames(data.tx) %in% c(taxdenom, outliers))]
data.tx_reads <- data.tx_reads[rowSums(data.tx_reads) != 0, ]

data.tx_tax <- data.tx[rownames(data.tx) %in% rownames(data.tx_reads), which(colnames(data.tx) %in% taxdenom)]

## PHYLOSEQ CREATION
OTU = otu_table(as.matrix(data.tx_reads), taxa_are_rows = TRUE)
TAX = tax_table(as.matrix(data.tx_tax))
SAM = sample_data(ids)

physeq = phyloseq(OTU, TAX, SAM)

if(isTRUE(savefiles)){
  save(data.tx, data.tx_reads, data.tx_tax, physeq, file = paste0(RSAV, "01_DATA_TAXA.rda"))
}

```

# KEGG DATASET PROCESSING:

```{r keggs_good, include=FALSE}
## DATASET PREPARATION
data.kg_an <- dplyr::select(data.kg, -one_of("Function", "Class", outliers))
data.kg_class <- data.kg[,c("Class", "Function"), drop = F]

# KEGGs accordint to methanogenesis (without Methane-related level):
data.kg_class$Association <- ifelse(str_detect(data.kg_class$Class, "Methane"), "Methane", "Others")

if(isTRUE(savefiles)){
  write.table(data.kg_class, paste0(WD, "Rsaves/KEGG_class.tsv"), sep="\t", dec=".", quote=F, col.names=NA)
}

# All classes in every KEGG...
nmax <- max(stringr::str_count(data.kg_class$Class, " \\| ")) + 1
## Divide up to 3rd class rank:
fun_sep <- separate(data.kg_class, Class, into = paste0("class_", 1:nmax), sep = ' \\| ') %>% 
  rownames_to_column("KEGG_ID") %>% 
  gather(key = "Class_ID", value = "Class", starts_with("class_")) %>%
  filter(!is.na(Class)) %>% filter(!(Class %in% as.character(0:100))) %>%
  separate(Class, into = paste0("Rank", 1:3), sep = '; ')
## Group by 2nd class rank:
fun_sep_2nd <- fun_sep %>% dplyr::select(-Rank3) %>% 
  group_by(KEGG_ID, Function, Rank1, Rank2) %>% summarise("Class_ID" = first(Class_ID))

# DEFINE TEXT COLUMNS FOR NEXT ANALYSIS:
textcols <- c("Class", "Function", "Rank1", "Rank2", "Rank3", "Association")
rankcols <- c("Rank1", "Rank2", "Rank3")

if(isTRUE(savefiles)){
  save(data.kg, data.kg_an, data.kg_class, fun_sep, textcols, rankcols, file = paste0(RSAV, "02_DATA_KEGG.rda"))
}

```

```{r keggs_GO, include=FALSE, eval=FALSE}
## KEGG TO GO (FOR REVIGO)
ko_go <- read.table(paste0(WD, "orthology_go.list"), sep = '\t', stringsAsFactors = F, header = F)
colnames(ko_go) <- c("KO", "GO", "links")
ko_go$KO <- str_replace(ko_go$KO, "ko:", "")
ko_go$GO <- str_replace(ko_go$GO, "go:", "GO:")
ko_go_filt <- ko_go[ko_go$KO %in% row.names(data.kg_class),]

sym_diff <- function(a,b) setdiff(union(a,b), intersect(a,b))
no_go <- sym_diff(unique(ko_go_filt$KO), row.names(data.kg_class))

# WITH ALL KEGGs:
#data.kg_meanra <- apply(data.kg_an, 2, function(x){x/sum(x)*100})
data.kg_meanra <- as.data.frame(apply(data.kg_an, 1, mean))
colnames(data.kg_meanra)[1] <- "MeanRA"

ko_go_fin <- merge(ko_go_filt, data.kg_meanra, by.x = 1, by.y = 0)
ko_go_fin <- ko_go_fin[, c(2, 4, 1, 3)]

#write.table(ko_go_fin, paste0(RSAV, "kegg_go.tsv"), sep = '\t', quote = F, row.names = F)

```

# Sample composition and Taxa prevalence

## Feature counts distribution in filtered final table

```{r count_distrib_taxa}
## PREVALENCE COMPOSITION IN TAXONOMY-FILTERED TABLE:
sgtp_mdata <- read.table(paste0(WD, "SQMreads/F2.tax_reads.tsv"), sep = '\t', stringsAsFactors = F, header = T) %>%
  dplyr::select(-one_of(taxdenom))
sgtp_mdata <- sgtp_mdata[rowSums(sgtp_mdata) != 0,]

## PREVALENCE COMPOSITION IN FINAL FILTERED TABLE:
counts_dataset.tx <- otu_table(physeq)@.Data

# SPARSITY BEFORE PREVALENCE FILTER:
(prod(dim(sgtp_mdata)) - Matrix::nnzero(sgtp_mdata))/prod(dim(sgtp_mdata))
# SPARSITY AFTER PREVALENCE FILTER:
(prod(dim(counts_dataset.tx)) - Matrix::nnzero(counts_dataset.tx))/prod(dim(counts_dataset.tx))
ggcounts <- count_distrib(counts_dataset.tx)
rm(sgtp_mdata)

```

```{r count_distrib_kegg}
## PREVALENCE COMPOSITION IN MULTINAME-FILTERED TABLE:
sgtp_mdata <- read.delim(paste0(WD, "SQMreads/F1.kegg_mreads.tsv"), stringsAsFactors=F, header=T, row.names=1) %>%
  dplyr::select(-one_of(c("Class", "Function")))
## PREVALENCE COMPOSITION IN FINAL FILTERED TABLE:
counts_dataset.kg <- data.kg_an

# SPARSITY BEFORE PREVALENCE FILTER:
(prod(dim(sgtp_mdata)) - Matrix::nnzero(sgtp_mdata))/prod(dim(sgtp_mdata))
# SPARSITY AFTER PREVALENCE FILTER:
(prod(dim(counts_dataset.kg)) - Matrix::nnzero(counts_dataset.kg))/prod(dim(counts_dataset.kg))
ggcounts_k <- count_distrib(counts_dataset.kg)
rm(sgtp_mdata)

```

## Taxa prevalence and relative abundance

```{r tax_prevalence_plot}
physeq_prev <- physeq
tax_table(physeq_prev)@.Data <- gsub("no \\w+ in NCBI", "no NCBI", tax_table(physeq_prev)@.Data)
tax_table(physeq_prev)@.Data <- gsub("Candidatus", "Cd.", tax_table(physeq_prev)@.Data)
colnames(tax_table(physeq_prev)@.Data) <- str_to_title(colnames(tax_table(physeq_prev)@.Data))

ggprev <- plot_taxa_prevalence(physeq_prev, prevalence.tax) +
  labs(title = "Taxa abundance & prevalence")

```

```{r tax_RA_plot}
# Need a Phyla_divisions.csv file with the desired information for new phyla classification (e.g., grouping low-abundance bacteria, rename no-NCBI eukaryotic phyla...)
# STRUCTURE OF Phyla_divisions.csv: 3 mandatory columns separated by ';'
## SK: Superkingdom
## PHY: Phylum as it appears in the phyloseq object (Candidatus, no-NCBI...)
## DIV: For clades with no-NCBI phylum entry, the new "Phylum" denomination wished.
# NOTE: we made a 5-column file. Modify it to make DIV column as wished. If barplot_phy needs changes, remember to match former classes (if needed).
phydiv <- read.csv(paste0(WD,"SQMReads/Phyla_divisions.csv"), sep = ";", stringsAsFactors = F)[,1:4]
phydiv[phydiv$DIV == "SAR", "DIV"] <- phydiv[phydiv$DIV == "SAR", "SDIV"]
phydiv[phydiv$SDIV == "Fungi", "DIV"] <- "Fungi"

## Use RA phyloseq object and group by 'plotlvl' (usually Phylum):
barplot_phy <- phyloseq_to_df(physeq) %>% 
  dplyr::select(-one_of("superkingdom", "class", "order", "family", "genus", "OTU")) %>%
  group_by(phylum) %>% summarise_all(sum) 
## Add DIV value in NULL cells according to the wished categories:
barplot_phy <- merge(phydiv[,-c(4)], barplot_phy, by.x = 2, by.y = 1)
barplot_phy[barplot_phy$SK == "Archaea", "DIV"] <- barplot_phy[barplot_phy$SK == "Archaea", "SK"]
major_bac <- c("Bacteroidetes", "Firmicutes", "Fibrobacteres")
barplot_phy[barplot_phy$PHY %in% major_bac, "DIV"] <- barplot_phy[barplot_phy$PHY %in% major_bac, "PHY"]
barplot_phy[barplot_phy$DIV == "", "DIV"] <- "Other Bacteria"
major_euk <- c("Fungi", "Hacrobia", "Stramenopiles", "Alveolata", "Rhizaria")
barplot_phy[barplot_phy$SK == "Eukaryota" & !(barplot_phy$DIV %in% major_euk), "DIV"] <- "Other Protozoa"
## Group table, calculate RA and prepare for plot (Sort samples by CH4):
barplot_phy <- dplyr::select(barplot_phy, -SK, -PHY) %>% group_by(DIV) %>% summarise_all(sum) %>% column_to_rownames("DIV")
barplot_phy <- apply(barplot_phy, 2, function(x){x/sum(x)*100})
barplot_phy <- data.frame(t(barplot_phy))
#colnames(barplot_phy) <- str_replace(colnames(barplot_phy), "..no.phylum.in.NCBI.", "_noNCBI")
barplot_phy <- merge(ids_out[,"ch4.spkpm_adjusted", drop = F], barplot_phy, by = 0)
colnames(barplot_phy)[c(1,2)] <- c("Sample", "CH4")
barplot_phy <- barplot_phy[order(barplot_phy$CH4),]
barplot_phy$Sample <- factor(barplot_phy$Sample, levels = barplot_phy$Sample)
barplot_phy <- pivot_longer(barplot_phy, -one_of("Sample", "CH4"), names_to = "Phylum", values_to = "Abundance")
barplot_phy$Phylum <- gsub("\\.", " ", barplot_phy$Phylum)
barplot_phy$Phylum <- factor(barplot_phy$Phylum, levels = c("Archaea", "Other Bacteria", "Bacteroidetes", "Firmicutes", "Fibrobacteres", "Fungi",
                                                            "Hacrobia", "Other Protozoa", "Stramenopiles", "Alveolata", "Rhizaria"))

# RA Plot:
raplot <- ggplot(barplot_phy, aes(x = Sample, y = Abundance, fill = Phylum)) + 
  geom_bar(stat = "identity", width = 1) + 
  scale_fill_manual(values = RColorBrewer::brewer.pal(11, "Set3")) +
  guides(fill = guide_legend(ncol =1)) +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(),
        legend.text = element_text(colour = "black", size = 10),
        legend.key.size = unit(.7, "cm"))

#ggplotly(raplot)

```

## Average relative abundance of features

```{r sunburst}
# SUNBURST: TAXONOMY
suna.tx <- apply(otu_table(physeq)@.Data, 2, function(x){x/sum(x)*100}) %>% apply(1, mean)
taxsun <- make_sunburst(sunreads = otu_table(physeq)@.Data, sunfeat = tax_table(physeq)@.Data, nranks = taxlvl, plottype = 1)

# SUNBURST: KEGG
## Prepare table (merge using fun_sep as base instead of with kegg_sep)
fun_unite <- unite(fun_sep, Class, Rank1, Rank2, Rank3, sep = "; ")
suna.kg <- merge(fun_unite, data.kg_an, by.x = "KEGG_ID", by.y = "row.names", all.x = T)
suna.kg[,-c(1:7)] <- apply(suna.kg[,-c(1:7)], 2, function(x){x/sum(x)})
sunb.kg <- data.frame("Class" = suna.kg$Class, "RA_mean" = rowMeans(suna.kg[,-c(1:7)])) %>% group_by(Class) %>% summarise_all(sum)
sunb.kg$Class <- gsub("; ", "-", gsub("-", "_", sunb.kg$Class))
keggsun <- sunburstR::sunburst(sunb.kg, legend = F, count = TRUE, width = "100%", height = 600)

```

## Genera proportions:

```{r genera_props}
piera <- apply(apply(otu_table(physeq)@.Data, 2, function(x){x/sum(x)*100}), 1, mean)

## PIE 1 TABLE: Unclassified vs RA > 1%: First, finish pieplot construction with 'pieplot$piera < 1'
pieplot <- merge(tax_table(physeq)@.Data[,c("phylum","genus")], as.data.frame(piera), by = 'row.names')
pieplot[,1:3] <- apply(pieplot[,1:3], 2, as.character)
pieplot$genus <- gsub("Unclassified .+", "Unclassified", pieplot$genus)
pieplot[pieplot$genus == "Unclassified", "phylum"] <- "Non-classified genera"
pieplot <- stats::aggregate(. ~ phylum + genus, pieplot[,-1], sum)

# Group genera with RA < 1%
pieplot[pieplot$piera < 1, "phylum"] <- "Genera with RA < 1%"
pieplot[pieplot$piera < 1, "genus"] <- "RA < 1%"
pieplot <- stats::aggregate(. ~ phylum + genus, pieplot, sum)
pieplot <- pieplot[order(pieplot$phylum, decreasing = T),]

piecols <- gg_color_hue(length(unique(pieplot$phylum)))
names(piecols) <- unique(pieplot$phylum)
pieplot$col <- piecols[match(pieplot$phylum, names(piecols))]
pieplot <-  remove_rownames(pieplot)

## PIE 2 TABLE: Only classified with RA > 0.1%: First, finish pieplot construction with 'pieplot$piera < 0.1'
pieplot2 <- merge(tax_table(physeq)@.Data[,c("phylum","genus")], as.data.frame(piera), by = 'row.names')
pieplot2[,1:3] <- apply(pieplot2[,1:3], 2, as.character)
pieplot2$genus <- gsub("Unclassified .+", "Unclassified", pieplot2$genus)
pieplot2[pieplot2$genus == "Unclassified", "phylum"] <- "Non-classified genera"
pieplot2 <- stats::aggregate(. ~ phylum + genus, pieplot2[,-1], sum)

# Group genera with RA < 0.1%
pieplot2[pieplot2$piera < 0.1, "phylum"] <- "Genera with RA < 0.1%"
pieplot2[pieplot2$piera < 0.1, "genus"] <- "RA < 0.1%"
pieplot2 <- stats::aggregate(. ~ phylum + genus, pieplot2, sum)
pieplot2 <- pieplot2[order(pieplot2$phylum, decreasing = T),]
pieplot2 <- pieplot2[pieplot2$genus != "Unclassified",]
pieplot2$piera100 <- pieplot2$piera*100/sum(pieplot2$piera)

piecols2 <- gg_color_hue(length(unique(pieplot2$phylum)))
names(piecols2) <- unique(pieplot2$phylum)
pieplot2$col <- piecols2[match(pieplot2$phylum, names(piecols2))]

## PIEPLOTS:
ggpie <- plot_ly() %>%
  add_pie(data = pieplot, labels = ~genus, values = ~piera, name = "Genus", textposition = 'auto', sort = F, 
          textinfo = 'label+percent', textfont = list(size = 14), marker = list(line = list(width = 1))) %>%
  layout(autosize = T, showlegend = F, colorway = pieplot$col,
         xaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE),
         yaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE))

ggpie2 <- plot_ly() %>%
  add_pie(data = pieplot2, labels = ~genus, values = ~piera100, name = "Genus", textinfo = 'label+percent', 
          textfont = list(size = 10), textposition = ifelse(pieplot2$piera100<1,"outside","inside"), 
          sort = F, marker = list(line = list(width = 1))) %>%
  layout(autosize = T, showlegend = F, colorway = pieplot2$col,
         xaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE),
         yaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE))

```

## Functionality proportions:

```{r func_props}
## DATA PREPARATION:
avgra.kg <- rowMeans(apply(data.kg_an, 2, function(x){(x/sum(x))*100}))
piekg <- merge(fun_sep, avgra.kg, by.x = "KEGG_ID", by.y = 0, all.x = T)
colnames(piekg)[8] <- "AvgRA"
piekg_path <- piekg[piekg$Rank1 != "Brite Hierarchies",]
piekg_brite <- piekg[piekg$Rank1 == "Brite Hierarchies",]

```

## REPORT: Save all plots:

```{r save_plots_1}
if(isTRUE(savefiles)){
  save(counts_distrib_tax_ggdat, ggcounts, physeq_prev, ggprev, phydiv, barplot_phy, raplot, suna.tx, taxsun, pieplot, pieplot2, 
	   piecols, piecols2, ggpie, ggpie2, file = paste0(RSAV, "11_GGPLOTS_TAXA.rda"))
  save(counts_distrib_kegg, ggcounts_k, sunb.kg, keggsun, piekg, piekg_path, piekg_brite, file = paste0(RSAV, "12_GGPLOTS_KEGG.rda"))
}

```

# Compositional Data Analysis

```{r CODA_ZCompositions_taxa, include=FALSE}
## DATA GROUP LIST
group_tax_all <- c("phylum", "class", "order", "family", "genus")
coda_data_tax <- phyloseq_to_df(physeq) %>% column_to_rownames("OTU")
coda_data_tax <- coda_data_tax[match(rownames(data.tx), rownames(coda_data_tax)),]
#coda_data_tax <- data[data$superkingdom == "Eukaryota",]

data_grps <- list()
for(i in seq(group_tax_all)){
  data_grps[[i]] <- list()
  groupranks <- taxdenom[1:which(taxdenom == group_tax_all[i])]
  nongroupranks <- taxdenom[-c(1:which(taxdenom == group_tax_all[i]))]
  grpdata <- dplyr::select(coda_data_tax, -c(colnames(coda_data_tax)[which(colnames(coda_data_tax) %in% nongroupranks)])) %>%
    group_by_at(groupranks) %>% summarise_all(sum)
  data_grps[[i]][[1]] <- grpdata[,!(colnames(grpdata) %in% head(groupranks, -1))] %>%
    column_to_rownames(var = group_tax_all[i])
  data_grps[[i]][[2]] <- grpdata[,groupranks]
  names(data_grps[[i]]) <- c("Counts", "Taxonomy")
  names(data_grps)[i] <- group_tax_all[i]
}

## ZCOMPOSITIONS WITH ZERO REPLACEMENT:
codata.GBM <- list(); codata.GBM.pro <- list(); codata.GBM.clr <- list(); codata.GBM.lra <- list()
tIn <- list(); pIn <- list(); ncmp <- list()
for(i in seq(data_grps)){
  GBMdataset <- data_grps[[i]]$Counts[,!(colnames(data_grps[[i]]$Counts) %in% outliers)] # MIGHT NOT WORK. CHECK.
  # Imputing zeros:
  codata.GBM[[i]] <- cmultRepl(t(GBMdataset), output="p-counts")
  names(codata.GBM)[i] <- names(data_grps)[i]
  # Compute RAs, at the same time transpose the matrix so that samples are in rows & genera in columns
  codata.GBM.pro[[i]] <- as.matrix(codata.GBM[[i]]) / rowSums(codata.GBM[[i]])
  names(codata.GBM.pro)[i] <- names(data_grps)[i]
  # CLR transformation
  codata.GBM.clr[[i]] <- easyCODA::CLR(codata.GBM.pro[[i]], weight = FALSE)
  names(codata.GBM.clr)[i] <- names(data_grps)[i]
  # LogRatios Analysis
  codata.GBM.lra[[i]] <- LRA(codata.GBM.pro[[i]], weight = FALSE)
  names(codata.GBM.lra)[i] <- names(data_grps)[i]
  # Total inertia == total variance; sv = singular values
  tIn[[i]] <- sum(codata.GBM.lra[[i]]$sv^2)
  names(tIn)[i] <- names(data_grps)[i]
  # percentages of inertia == variance % explained by each component
  # sum of all the % on every component = 100
  pIn[[i]] <- 100 * codata.GBM.lra[[i]]$sv^2 / tIn[[i]]
  names(pIn)[i] <- names(data_grps)[i]
  # total number of components = n_samples - 1
  ncmp[[i]] <- length(pIn[[i]])
  names(ncmp)[i] <- names(data_grps)[i]
}

## REPORT: Save files.
if(isTRUE(savefiles)){
  save(coda_data_tax, data_grps, codata.GBM, codata.GBM.pro, codata.GBM.clr, codata.GBM.lra, tIn, pIn, ncmp, 
       file = paste0(RSAV, "21_CODATA_TAXA.rda"))
}

```

```{r CODA_ZCompositions_kegg, include=FALSE}
## ZCOMPOSITIONS WITH ZERO REPLACEMENT (remove outlier samples):
GBMdataset_k <- data.kg_an[,!(colnames(data.kg_an) %in% outliers)]
# Imputing zeros:
codata.GBM_k <- cmultRepl(t(GBMdataset_k), output="p-counts")
# Compute RAs, at the same time transpose the matrix so that samples are in rows & genera in columns
codata.GBM.pro_k <- as.matrix(codata.GBM_k) / rowSums(codata.GBM_k)
# CLR transformation
codata.GBM.clr_k <- easyCODA::CLR(codata.GBM.pro_k, weight = FALSE)
# LogRatios Analysis (unweighted LRA = PCA of CLR; LRA = weighted PCA of CLR)
codata.GBM.lra_k <- LRA(codata.GBM.pro_k, weight = FALSE)
# Total inertia == total variance; sv = singular values
tIn_k <- sum(codata.GBM.lra_k$sv^2)
# percentages of inertia == variance % explained by each component
# sum of all the % on every component = 100
pIn_k <- 100 * codata.GBM.lra_k$sv^2 / tIn_k
# total number of components = n_samples - 1
ncmp_k <- length(pIn_k)

#plot(codata.GBM.lra_k)

## REPORT: Save files.
if(isTRUE(savefiles)){
  save(GBMdataset_k, codata.GBM_k, codata.GBM.pro_k, codata.GBM.clr_k, codata.GBM.lra_k, tIn_k, pIn_k, ncmp_k,
       file = paste0(RSAV, "22_CODATA_KEGG.rda"))
}

```

# Alpha-diversity

```{r alphadata, include=FALSE} 
measures <- c("Shannon", "InvSimpson")
phyalpha <- physeq

alpharich <- estimate_richness(phyalpha, measures = measures)
alphadata <- cbind(sample_data(phyalpha), alpharich)
alphadata$SSum <- sample_sums(phyalpha)

indvars <- c("CH4_Rank", "NLACTA", "DIASLE")

# Test normality & homocedasticity for independent variables (indvars) and make group contrast on dependent
# variables (measures) depending on the results.
normvartest <- list(); densp <- list()
for(i in seq(indvars)){
  normvartest[[i]] <- Test_oneway(alphadata, indvars[i], measures)
  densp[[i]] <- list()
  for(m in seq(measures)){
    indvsmeas <- na.omit(alphadata[,c(indvars[i], measures[m])])
    densp[[i]][[m]] <- ggplot(indvsmeas, aes_string(x=measures[m], fill=indvars[i])) + 
      geom_density(alpha = 0.5) + xlab(measures[m]) +
      theme(axis.title.x = element_text(colour=ifelse(measures[m] %in% names(normvartest[[i]]$Non_Parametric),'red','black')))
  }
  names(normvartest)[i] <- indvars[i]
  names(densp)[i] <- indvars[i]
}

## Alpha-diversity ANOVA:
richness.anova <- list()
for (i in seq(measures)){
  richness.anova[[i]] <- aov(alphadata[,measures[i]] ~ NLACTA + DIASLE * CH4_Rank, alphadata)
  names(richness.anova)[i] <- measures[i]
}

lapply(richness.anova, summary)
interaction.plot(alphadata$CH4_Rank, alphadata$DIASLE, alphadata$Shannon)

# We apply type III sum of squares, as significant interaction has been observed (Only for DIASLE).
lapply(richness.anova, car::Anova, type="III")

## REPORT: Save files.
if(isTRUE(savefiles)){
  save(alphadata, normvartest, densp, richness.anova, file = paste0(RSAV, "30_DIV_ALPHA.rda"))
}

```

# Beta-diversity
## Principal Component Analysis (PCA)

```{r betadiv, include=FALSE}
## Phenotype data for beta-diversity. We remove outlier samples (SH_3, AMET_016).
bphenots <- ids_out[!(rownames(ids_out) %in% outliers),]
bphenots$CH4_Rank <- cut(bphenots$ch4.spkpm_adjusted, 
                         breaks = c(quantile(bphenots$ch4.spkpm_adjusted, prob = c(0, 0.25, 0.50, 0.75, 1))),
                         labels=c("LOW","L_MID","H_MID","HIGH"), include.lowest = TRUE)
bphenots <- dplyr::select(bphenots, one_of(phenots))
renmphenots <- c("CH4", "NL", "DM", "CH4_Rank")

## Dataset 1: taxonomy:
betadat.lst <- list()
for (i in seq(codata.GBM.clr)){
  betadat <- codata.GBM.clr[[i]]$LR
  betadat.lst[[i]] <- merge(bphenots, as.data.frame(betadat), by = "row.names") %>% column_to_rownames("Row.names")
  betadat.lst[[i]] <- betadat.lst[[i]][!(rownames(betadat.lst[[i]]) %in% outliers),]
  colnames(betadat.lst[[i]]) <- gsub("no \\w+ in NCBI", "no NCBI", colnames(betadat.lst[[i]]))
  colnames(betadat.lst[[i]]) <- gsub("Candidatus", "Cd.", colnames(betadat.lst[[i]]))
  colnames(betadat.lst[[i]])[1:length(bphenots)] <- renmphenots
  names(betadat.lst)[i] <- names(codata.GBM.clr)[i]
}
# PCA 1: Taxonomy:
beta.pca <- list(); enft <- list()
for(i in seq(betadat.lst)){
  beta.pca[[i]] <- prcomp(betadat.lst[[i]] %>% dplyr::select(-one_of(renmphenots)))
  names(beta.pca)[i] <- names(betadat.lst)[i]
  #stressplot(beta.mds)
  set.seed(17)
  enft[[i]] <- envfit(beta.pca[[i]] ~ NL + DM + CH4, data = betadat.lst[[i]], permutations = 999)
  names(enft)[i] <- names(betadat.lst)[i]
}
# PCA plot 1: Taxonomy:
beta.p <- list(); beta.p.layer <- list()
for(i in seq(beta.pca)){
  beta.p[[i]] <- ggpca_fit(beta.pca[[i]], betadat.lst[[i]][,1:4], c("CH4", "NL", "DM"), "CH4_Rank")
  names(beta.p)[i] <- names(beta.pca)[i]
  beta.p.layer[[i]] <- list(labs(title = paste("PCA -", names(beta.p)[i], "level")), theme_classic())
  names(beta.p.layer)[i] <- names(beta.pca)[i]
}

## Dataset 2: KEGGs:
betadat_k <- codata.GBM.clr_k$LR
betadat_k <- merge(bphenots, as.data.frame(betadat_k), by = "row.names") %>% column_to_rownames("Row.names")
colnames(betadat_k)[1:length(bphenots)] <- renmphenots
# PCA 2: KEGGs:
beta.pca.k <- prcomp(betadat_k %>% dplyr::select(-one_of(renmphenots)))
set.seed(17)
enft.k <- envfit(beta.pca.k ~ NL + DM + CH4, data = betadat_k, permutations = 999)
# PCA plot 2: KEGGs:
beta.p.k <- ggpca_fit(beta.pca.k, betadat_k[,1:4], c("CH4", "NL", "DM"), "CH4_Rank")
beta.p.k.layer <- list(labs(title = "PCA - KEGGs"), theme_classic())

## REPORT: Save files.
if(isTRUE(savefiles)){
  save(bphenots, renmphenots, betadat.lst, beta.pca, enft, beta.p, beta.p.layer, file = paste0(RSAV, "31_DIV_BETA_TAXA.rda"))
  save(betadat_k, beta.pca.k, enft.k, beta.p.k, beta.p.k.layer, file = paste0(RSAV, "32_DIV_BETA_KEGG.rda"))
}

```

## Permutational Multivariate analysis of variance (PERMANOVA)

```{r beta_permanova}
## PERMANOVA ANALYSIS: Taxonomy:
adonis.lst <- list()
for(i in seq(betadat.lst)){
  betadist <- dist(betadat.lst[[i]] %>% dplyr::select(-one_of(renmphenots)), method = "euclidean")
  adonis.lst[[i]] <- adonis(betadist ~ DM + NL + CH4_Rank, data = betadat.lst[[i]], perm = 999)
  names(adonis.lst)[i] <- names(betadat.lst)[i]
}

## PERMANOVA ANALYSIS: KEGGs:
betadist_k <- dist(betadat_k %>% dplyr::select(-one_of(renmphenots)), method = "euclidean")
adonis.k <- adonis(betadist_k ~ DM + NL + CH4_Rank, data = betadat_k, perm = 999)

## REPORT: Save files.
if(isTRUE(savefiles)){
  save(adonis.lst, file = paste0(RSAV, "33_DIV_BETA_PERMANOVA_TAXA.rda"))
  save(adonis.k, file = paste0(RSAV, "34_DIV_BETA_PERMANOVA_KEGG.rda"))
}

```

## Ordisurf: Unmasking CH4 emissions effect

```{r betadiv_ordisurf, include=FALSE}
# We use PCA to make an ordisurf plot comparing linear fitting vs GAM fitting of CH4 (residuals from NL+DM correction).

## ORDISURF DATA 1: Taxonomy:
ordisurf.lst <- list()
for(i in seq(beta.pca)){
  ch4.adj <- lm(CH4 ~ NL + DM, data = betadat.lst[[i]])
  betadat.lst[[i]] <- add_column(betadat.lst[[i]], CH4.res = residuals(ch4.adj), .after = 4)
  ordisurf.lst[[i]] <- list()
  ordisurf.lst[[i]]$LM <- ordisurf(beta.pca[[i]], betadat.lst[[i]]$CH4, plot = F, scaling = 2, knots = 1)
  ordisurf.lst[[i]]$LM.res <- ordisurf(beta.pca[[i]], betadat.lst[[i]]$CH4.res, plot = F, scaling = 2, knots = 1)
  ordisurf.lst[[i]]$GAM <- ordisurf(beta.pca[[i]], betadat.lst[[i]]$CH4, plot = F, scaling = 2)
  ordisurf.lst[[i]]$GAM.res <- ordisurf(beta.pca[[i]], betadat.lst[[i]]$CH4.res, plot = F, scaling = 2)
  names(ordisurf.lst)[i] <- names(beta.pca)[i]
}

## ORDISURF DATA 2: KEGGs:
ch4.adj <- lm(CH4 ~ NL + DM, data = betadat_k)
betadat_k <- add_column(betadat_k, CH4.res = residuals(ch4.adj), .after = 4)
ordisurf.k <- list()
ordisurf.k$LM <- ordisurf(beta.pca.k, betadat_k$CH4, plot = F, scaling = 2, knots = 1)
ordisurf.k$LM.res <- ordisurf(beta.pca.k, betadat_k$CH4.res, plot = F, scaling = 2, knots = 1)
ordisurf.k$GAM <- ordisurf(beta.pca.k, betadat_k$CH4, plot = F, scaling = 2)
ordisurf.k$GAM.res <- ordisurf(beta.pca.k, betadat_k$CH4.res, plot = F, scaling = 2)

## REPORT: Save files.
if(isTRUE(savefiles)){
  save(ordisurf.lst, file = paste0(RSAV, "35_DIV_BETA_PCA_ORDISURF_TAXA.rda"))
  save(ordisurf.k, file = paste0(RSAV, "36_DIV_BETA_PCA_ORDISURF_KEGG.rda"))
}

```

# Differential Abundance
## LIMMA

```{r limma_CLR, include=FALSE}
## Get OTU/KEGG abundances and sample metadata:
meta <- ids
otu <- t(codata.GBM.clr$genus$LR)
kegg <- t(codata.GBM.clr_k$LR)

## LIMMA 1: Taxonomy:
# Limma design:
design_t <- model.matrix(~ 0 + CH4_Rank, data = meta)
colnames(design_t) <- levels(factor(meta$CH4_Rank)) %>% str_replace(" - ", ".")
# Fit the limma model & make contrasts (in the subtraction, 2nd component will be treated as reference level)
fit.pre.t <- lmFit(otu, design_t)
cont.matrix.t <- makeContrasts(LOWvsHIGH = HIGH-LOW, LOWvsL_MID = L_MID-LOW, LOWvsH_MID = H_MID-LOW, HIGHvsH_MID = HIGH-H_MID, levels = design_t)
fit.tax <- contrasts.fit(fit.pre.t, cont.matrix.t)
fit.tax <- eBayes(fit.tax)
# Results:
# Down-abundant feats (<0): More associated to reference level (2nd contrast component)
# Up-abundant feats (>0): More associated to "treatment" level (1st contrast component)
results.tax <- decideTests(fit.tax, p.value = qval, lfc = fc.cut)
summary(results.tax)
# TopTables:
toptbl <- list()
for(i in seq(ncol(cont.matrix.t))){
  toptbl[[i]] <- topTable(fit.tax, coef = i, p.value = qval, lfc = fc.cut, number = 'Inf')
  names(toptbl)[i] <- colnames(cont.matrix.t)[i]
}

## LIMMA 2: KEGGs:
# Limma design:
design_k <- model.matrix(~ 0 + CH4_Rank, data = meta)
colnames(design_k) <- levels(factor(meta$CH4_Rank)) %>% str_replace(" - ", ".")
# Fit the limma model & Make contrasts (in the subtraction, 2nd component will be trated as reference level)
fit.pre.k <- lmFit(kegg, design_k)
cont.matrix.k <- makeContrasts(LOWvsHIGH = HIGH-LOW, LOWvsL_MID = L_MID-LOW, LOWvsH_MID = H_MID-LOW, L_MIDvsH_MID = H_MID-L_MID,
                               HIGHvsH_MID = HIGH-H_MID, HIGHvsL_MID = HIGH-L_MID, levels = design_k)
fit.kegg <- contrasts.fit(fit.pre.k, cont.matrix.k)
fit.kegg <- eBayes(fit.kegg)
# Results:
# Down-abundant feats (<0): More associated to reference level (2nd contrast component)
# Up-abundant feats (>0): More associated to "treatment" level (1st contrast component)
results.kegg <- decideTests(fit.kegg, p.value = qval, lfc = fc.cut)
summary(results.kegg)
# TopTables:
toptbl_k <- list()
for(i in seq(ncol(cont.matrix.k))){
  toptbl_k[[i]] <- topTable(fit.kegg, coef = i, p.value = qval, lfc = fc.cut, number = 'Inf')
  names(toptbl_k)[i] <- colnames(cont.matrix.k)[i]
}


## REPORT: Save files.
if(isTRUE(savefiles)){
  save(design_t, cont.matrix.t, fit.tax, results.tax, toptbl, file = paste0(RSAV, "41_LIMMA_TAXA.rda"))
  save(design_k, cont.matrix.k, fit.kegg, results.kegg, toptbl_k, file = paste0(RSAV, "42_LIMMA_KEGG.rda"))
}

## CUMULATED RA OF LOW AND HIGH DA TAXA/KEGGS:
# Taxa:
genra <- apply(apply(otu_table(physeq)@.Data, 2, function(x){x/sum(x)*100}), 1, mean)
gentax <- tax_table(physeq)@.Data[,"genus"]
genav <- data.frame("RA"=genra, "Genus"=gentax, stringsAsFactors = F)

HIGHOA <- unique(unlist(lapply(toptbl[c("LOWvsHIGH","LOWvsH_MID")], function(x){rownames(x[x$logFC > 0,])})))
LOWOA <- unique(unlist(lapply(toptbl[c("LOWvsHIGH","LOWvsH_MID")], function(x){rownames(x[x$logFC < 0,])})))

sum(genav[genav$Genus %in% HIGHOA,"RA"])
sum(genav[genav$Genus %in% LOWOA,"RA"])

# KEGGs:
kegra <- apply(apply(data.kg_an, 2, function(x){x/sum(x)*100}), 1, mean)
kegav <- data.frame("RA"=kegra, "KEGG"=names(kegra), stringsAsFactors = F)

HIGHOA <- unique(unlist(lapply(toptbl_k[c("LOWvsHIGH","LOWvsH_MID")], function(x){rownames(x[x$logFC > 0,])})))
LOWOA <- unique(unlist(lapply(toptbl_k[c("LOWvsHIGH","LOWvsH_MID")], function(x){rownames(x[x$logFC < 0,])})))

sum(kegav[kegav$KEGG %in% HIGHOA,"RA"])
sum(kegav[kegav$KEGG %in% LOWOA,"RA"])

```


## KEGG TO TAXA
```{r kegg_tax_tableprep, include=FALSE, eval=FALSE}
# TABLE PREPARATION:
#Use Kegg_class.tsv (data.kg_class) and DA KEGG analysis to associate Keggs to methane emissions:
#CH4-related group will be formed by DA KEGGs not directly involved in methanogenesis.
## Load Kegg_class.tsv:
data.kg_class <- read.delim(paste0(WD,"Rsaves/Kegg_class.tsv"))
## Load DA KEGG names (replace "Others" to "Methane-related as now we are considering every DA KEGG as methane-related):
DAkeggs <- openxlsx::read.xlsx(paste0(WD, "Rsaves/DA_Limma_paper.xlsx"), sheet = 2)
DAkeggs$Association[DAkeggs$Association == "Others"] <- "Methane-related"
DAkeggs_LOW <- unique(DAkeggs[DAkeggs$logFC < 0,"KEGG"])
DAkeggs_HIGH <- unique(DAkeggs[DAkeggs$logFC > 0,"KEGG"])
DAkeggs_methrel <- unique(DAkeggs$KEGG[DAkeggs$Association != "Methane"])
## Replace Associations in data.kg_class:
levels(data.kg_class$Association) <- c(levels(data.kg_class$Association), "Methane-related")
data.kg_class$Association[data.kg_class$X %in% DAkeggs_methrel] <- "Methane-related"
## Subset data.kg_class only with DA keggs:
data.kg_class_LOW <- data.kg_class[data.kg_class$X %in% DAkeggs_LOW,] %>% remove_rownames %>% column_to_rownames("X")
data.kg_class_HIGH <- data.kg_class[data.kg_class$X %in% DAkeggs_HIGH,] %>% remove_rownames %>% column_to_rownames("X")
write.table(data.kg_class_LOW, paste0(WD, "Rsaves/KEGG_class_LOW.tsv"), sep="\t", dec=".", quote=F, col.names=NA)
write.table(data.kg_class_HIGH, paste0(WD, "Rsaves/KEGG_class_HIGH.tsv"), sep="\t", dec=".", quote=F, col.names=NA)

```

```{r kegg_tax_corresp, include=FALSE}
# VARS FOR SUPERIOR TAXONOMY RANKS:
taxdenom_init <- c("k", substring(taxdenom[-1], 1, 1))
## Load species taxonomy results table:
data.tx_sp <- read.delim(paste0(WD, "SQMreads/METALGEN4sub_taxonomy_new_new/", "combined.species.allfilter.abund.tsv"),
                         stringsAsFactors = F)[,1, drop = F] %>% separate(X, taxdenom, sep = ";")
data.tx_sp <- as.data.frame(apply(data.tx_sp, 2, gsub, pattern = "\\w_", replacement = ""), stringsAsFactors = F)
## Animal and plant phyla:
nope <- c("Acanthocephala", "Annelida", "Arthropoda", "Brachiopoda", "Bryozoa", "Chaetognatha", "Chlorophyta", "Chordata",
          "Cnidaria", "Ctenophora", "Echinodermata", "Entoprocta", "Gnathostomulida", "Hemichordata", "Mollusca", "Nematoda",
          "Nematomorpha", "Nemertea", "Onychophora", "Orthonectida", "Placozoa", "Platyhelminthes", "Porifera", "Rhodophyta",
          "Rotifera", "Streptophyta", "Tardigrada", "Xenacoelomorpha")
## Load Eukaryote divisions:
phydiv <- read.csv(paste0(WD,"SQMReads/Phyla_divisions.csv"), sep = ";", stringsAsFactors = F)[,1:4]
#phydiv[phydiv$DIV == "Metamonada", "SDIV"] <- phydiv[phydiv$DIV == "Metamonada", "DIV"]
#phydiv[phydiv$DIV == "Amoebozoa", "SDIV"] <- phydiv[phydiv$DIV == "Amoebozoa", "DIV"]

# METHANOGENESIS-RELATED KEGGs (DA KEGGs): LOW-ASSOCIATED: (Reads_MRelLOW.tsv generated in CESGA from SQMr_exe.sh)
rds_mrel_L <- read.table(paste0(WD,"SQMreads/Reads_MRelLOW.tsv"), sep = "\t", stringsAsFactors = F, row.names = 2)
colnames(rds_mrel_L) <- c("Sample", "Taxonomy", "KEGG")
rds_mrel_L <- separate(rds_mrel_L, Taxonomy, into = c("Rank", "Taxon"), sep = "_")
rds_mrel_L[is.na(rds_mrel_L)] <- ""
rds_mrel_L$KEGG <- gsub(rds_mrel_L$KEGG, pattern = "\\*", replacement = "")
rds_mrel_L$Assoc <- "LOW"

# METHANOGENESIS-RELATED KEGGs (DA KEGGs): HIGH-ASSOCIATED: (Reads_MRelHIGH.tsv generated in CESGA from SQMr_exe.sh)
rds_mrel_H <- read.table(paste0(WD,"SQMreads/Reads_MRelHIGH.tsv"), sep = "\t", stringsAsFactors = F, row.names = 2)
colnames(rds_mrel_H) <- c("Sample", "Taxonomy", "KEGG")
rds_mrel_H <- separate(rds_mrel_H, Taxonomy, into = c("Rank", "Taxon"), sep = "_")
rds_mrel_H[is.na(rds_mrel_H)] <- ""
rds_mrel_H$KEGG <- gsub(rds_mrel_H$KEGG, pattern = "\\*", replacement = "")
rds_mrel_H$Assoc <- "HIGH"

# BIND BOTH:
rds_mrel <- rbind(rds_mrel_L, rds_mrel_H)

## ADD SUPERIOR TAXONOMY RANKS
### Merge taxonomy with Eukaryotes divisions and remove animal and plant taxa:
data.tx_sp_2 <- merge(data.tx_sp, phydiv[,-1], by.x = 2, by.y = 1, sort = F, all.x = T) %>% rename(division = DIV) %>%
  relocate(division, SDIV, phylum, .after = superkingdom) %>% na_if("")
data.tx_sp_2$division <- ifelse(data.tx_sp_2$superkingdom != "Eukaryota", data.tx_sp_2$phylum, data.tx_sp_2$division)
data.tx_sp_2$SDIV <- ifelse(data.tx_sp_2$superkingdom != "Eukaryota", data.tx_sp_2$division, data.tx_sp_2$SDIV)
data.tx_sp_2$SDIV <- ifelse(is.na(data.tx_sp_2$SDIV) & !is.na(data.tx_sp_2$division), 
                            data.tx_sp_2$division, data.tx_sp_2$SDIV)

### Remove reads without taxonomy or with classification out of the 7 principal taxonomic ranks:
# 16.9% of reads assigned to DA KEGGs were not assigned to taxonomy.
rds_mrel_tax <- rds_mrel[!(rds_mrel$Rank %in% c("", "n")),]

### Add superkingdom and phylum classifications to each read:
rds_mrel_merged <- NULL
for(i in taxdenom_init){
  # X: contains reads classified up to [i] taxonomic level
  X <- rds_mrel_tax[rds_mrel_tax$Rank == i,] %>% rownames_to_column("Read")
  # Y: taxonomy matrix with superkingdom, phylum and key taxonomic rank (when needed)
  if (i == "k"){ Y <- distinct(data.tx_sp_2[,"superkingdom", drop = F])
  } else if (i == "p") { Y <- distinct(data.tx_sp_2[,unique(c("superkingdom", "division", "SDIV", "phylum"))])
  } else { Y <- distinct(data.tx_sp_2[,unique(c("superkingdom", "division", "SDIV", "phylum", 
                                                taxdenom[match(i, taxdenom_init)]))]) }
  # Merged X+Y:
  merged <- merge(X, Y, by.x = "Taxon", by.y = taxdenom[match(i, taxdenom_init)], all.x = F)
  # Sort columns for matching in rbind:
  merged <- merged[,c(2:4,1,5:ncol(merged))]
  if(i == "k"){
    merged$superkingdom <- merged$Taxon
    merged$division <- NA; merged$SDIV <- NA; merged$phylum <- NA
  } else if (i == "p") { 
    merged$phylum <- merged$Taxon
  }
  # Bind all [i] merged tables:
  rds_mrel_merged <- rbind(rds_mrel_merged, merged)
  # Reset variables:
  rm(X, Y, merged)
}

### Remove virus, animal and plant taxa (OR add DIV and SDIV):
nonmicstate <- "remove"
if (nonmicstate == "remove"){
  rds_mrel_merged <- rds_mrel_merged[!(rds_mrel_merged$superkingdom == "Viruses"),]
  rds_mrel_merged <- rds_mrel_merged[!(rds_mrel_merged$phylum %in% nope),]
  ### Check remaining NA divisions and remove animal ones
  unique(rds_mrel_merged[rds_mrel_merged$superkingdom == "Eukaryota" & !is.na(rds_mrel_merged$phylum) &
                           is.na(rds_mrel_merged$division),"phylum"])
  rds_mrel_merged <- rds_mrel_merged[-which(str_detect(rds_mrel_merged$phylum, "Priapulimorpha|Rhopaluridae")),]
} else if (nonmicstate == "add"){
  rds_mrel_merged[which(rds_mrel_merged$phylum == "Streptophyta"), "division"] <- "Plantae"
  rds_mrel_merged[which(rds_mrel_merged$phylum == "Streptophyta"), "SDIV"] <- "Viridiplantae"
  rds_mrel_merged[rds_mrel_merged$phylum %in% nope[!(nope=="Streptophyta")], "division"] <- "Opisthokonta"
  rds_mrel_merged[rds_mrel_merged$phylum %in% nope[!(nope=="Streptophyta")], "SDIV"] <- "Animalia"
  ### Check remaining NA divisions and remove animal ones
  unique(rds_mrel_merged[rds_mrel_merged$superkingdom == "Eukaryota" & !is.na(rds_mrel_merged$phylum) &
                           is.na(rds_mrel_merged$division),"phylum"])
  rds_mrel_merged[which(str_detect(rds_mrel_merged$phylum, "Priapulimorpha|Rhopaluridae")), "division"] <- "Opisthokonta"
  rds_mrel_merged[which(str_detect(rds_mrel_merged$phylum, "Priapulimorpha|Rhopaluridae")), "SDIV"] <- "Animalia"
} else {print("ERROR, nonmicstate must be 'remove' or 'add'")}

### Check remaining NA divisions
rds_mrel_merged[which(rds_mrel_merged$phylum == "fungal sp. (no phylum in NCBI)"), "division"] <- "Opisthokonta"
rds_mrel_merged[which(rds_mrel_merged$phylum == "fungal sp. (no phylum in NCBI)"), "SDIV"] <- "Fungi"

## REPORT: Save files.
if(isTRUE(savefiles)){
  save(rds_mrel_merged, file = paste0(RSAV, "43_DA_KEGG_TAXA.rda"))
}

```

```{r kegg_tax_tableprep_CH4, include=FALSE, eval=FALSE}
# TABLE PREPARATION:
#Use Kegg_class.tsv (data.kg_class) to take KEGGs from methanogenesis pathway (CH4 KEGGs):
## Load Kegg_class.tsv:
data.kg_class <- read.delim(paste0(WD,"Rsaves/Kegg_class.tsv"))
CH4keggs <- data.kg_class[data.kg_class$Association == "Methane",] %>% remove_rownames %>% column_to_rownames("X")
## Subset data.kg_class only with DA keggs:
write.table(CH4keggs, paste0(WD, "Rsaves/KEGG_class_CH4.tsv"), sep="\t", dec=".", quote=F, col.names=NA)

```

```{r kegg_tax_corresp_CH4, include=FALSE}
# VARS FOR SUPERIOR TAXONOMY RANKS:
taxdenom_init <- c("k", substring(taxdenom[-1], 1, 1))
## Load species taxonomy results table:
data.tx_sp <- read.delim(paste0(WD, "SQMreads/METALGEN4sub_taxonomy_new_new/", "combined.species.allfilter.abund.tsv"),
                         stringsAsFactors = F)[,1, drop = F] %>% separate(X, taxdenom, sep = ";")
data.tx_sp <- as.data.frame(apply(data.tx_sp, 2, gsub, pattern = "\\w_", replacement = ""), stringsAsFactors = F)
## Animal and plant phyla:
nope <- c("Acanthocephala", "Annelida", "Arthropoda", "Brachiopoda", "Bryozoa", "Chaetognatha", "Chlorophyta", "Chordata",
          "Cnidaria", "Ctenophora", "Echinodermata", "Entoprocta", "Gnathostomulida", "Hemichordata", "Mollusca", "Nematoda",
          "Nematomorpha", "Nemertea", "Onychophora", "Orthonectida", "Placozoa", "Platyhelminthes", "Porifera", "Rhodophyta",
          "Rotifera", "Streptophyta", "Tardigrada", "Xenacoelomorpha")
## Load Eukaryote divisions:
phydiv <- read.csv(paste0(WD,"SQMReads/Phyla_divisions.csv"), sep = ";", stringsAsFactors = F)[,1:4]
#phydiv[phydiv$DIV == "Metamonada", "SDIV"] <- phydiv[phydiv$DIV == "Metamonada", "DIV"]
#phydiv[phydiv$DIV == "Amoebozoa", "SDIV"] <- phydiv[phydiv$DIV == "Amoebozoa", "DIV"]

# METHANOGENESIS KEGGs: (Reads_CH4.tsv generated in CESGA from SQMr_exe.sh)
rds_mrel <- read.table(paste0(WD,"SQMreads/Reads_CH4.tsv"), sep = "\t", stringsAsFactors = F, row.names = 2)
colnames(rds_mrel) <- c("Sample", "Taxonomy", "KEGG")
rds_mrel <- separate(rds_mrel, Taxonomy, into = c("Rank", "Taxon"), sep = "_")
rds_mrel[is.na(rds_mrel)] <- ""
rds_mrel$KEGG <- gsub(rds_mrel$KEGG, pattern = "\\*", replacement = "")

## ADD SUPERIOR TAXONOMY RANKS
### Merge taxonomy with Eukaryotes divisions and remove animal and plant taxa:
data.tx_sp_2 <- merge(data.tx_sp, phydiv[,-1], by.x = 2, by.y = 1, sort = F, all.x = T) %>% rename(division = DIV) %>%
  relocate(division, SDIV, phylum, .after = superkingdom) %>% na_if("")
data.tx_sp_2$division <- ifelse(data.tx_sp_2$superkingdom != "Eukaryota", data.tx_sp_2$phylum, data.tx_sp_2$division)
data.tx_sp_2$SDIV <- ifelse(data.tx_sp_2$superkingdom != "Eukaryota", data.tx_sp_2$division, data.tx_sp_2$SDIV)
data.tx_sp_2$SDIV <- ifelse(is.na(data.tx_sp_2$SDIV) & !is.na(data.tx_sp_2$division), 
                            data.tx_sp_2$division, data.tx_sp_2$SDIV)

### Remove reads without taxonomy or with classification out of the 7 principal taxonomic ranks:
# 15.8% of reads assigned to DA KEGGs were not assigned to taxonomy.
rds_mrel_tax <- rds_mrel[!(rds_mrel$Rank %in% c("", "n")),]

### Add superkingdom and phylum classifications to each read:
rds_mrel_merged <- NULL
for(i in taxdenom_init){
  # X: contains reads classified up to [i] taxonomic level
  X <- rds_mrel_tax[rds_mrel_tax$Rank == i,] %>% rownames_to_column("Read")
  # Y: taxonomy matrix with superkingdom, phylum and key taxonomic rank (when needed)
  if (i == "k"){ Y <- distinct(data.tx_sp_2[,"superkingdom", drop = F])
  } else if (i == "p") { Y <- distinct(data.tx_sp_2[,unique(c("superkingdom", "division", "SDIV", "phylum"))])
  } else { Y <- distinct(data.tx_sp_2[,unique(c("superkingdom", "division", "SDIV", "phylum", 
                                                taxdenom[match(i, taxdenom_init)]))]) }
  # Merged X+Y:
  merged <- merge(X, Y, by.x = "Taxon", by.y = taxdenom[match(i, taxdenom_init)], all.x = F)
  # Sort columns for matching in rbind:
  merged <- merged[,c(2:4,1,5:ncol(merged))]
  if(i == "k"){
    merged$superkingdom <- merged$Taxon
    merged$division <- NA; merged$SDIV <- NA; merged$phylum <- NA
  } else if (i == "p") { 
    merged$phylum <- merged$Taxon
  }
  # Bind all [i] merged tables:
  rds_mrel_merged <- rbind(rds_mrel_merged, merged)
  # Reset variables:
  rm(X, Y, merged)
}

### Remove virus, animal and plant taxa (OR add DIV and SDIV):
nonmicstate <- "remove"
if (nonmicstate == "remove"){
  rds_mrel_merged <- rds_mrel_merged[!(rds_mrel_merged$superkingdom == "Viruses"),]
  rds_mrel_merged <- rds_mrel_merged[!(rds_mrel_merged$phylum %in% nope),]
  ### Check remaining NA divisions and remove animal ones
  unique(rds_mrel_merged[rds_mrel_merged$superkingdom == "Eukaryota" & !is.na(rds_mrel_merged$phylum) &
                           is.na(rds_mrel_merged$division),"phylum"])
} else if (nonmicstate == "add"){
  rds_mrel_merged[which(rds_mrel_merged$phylum == "Streptophyta"), "division"] <- "Plantae"
  rds_mrel_merged[which(rds_mrel_merged$phylum == "Streptophyta"), "SDIV"] <- "Viridiplantae"
  rds_mrel_merged[rds_mrel_merged$phylum %in% nope[!(nope=="Streptophyta")], "division"] <- "Opisthokonta"
  rds_mrel_merged[rds_mrel_merged$phylum %in% nope[!(nope=="Streptophyta")], "SDIV"] <- "Animalia"
  ### Check remaining NA divisions and remove animal ones
  unique(rds_mrel_merged[rds_mrel_merged$superkingdom == "Eukaryota" & !is.na(rds_mrel_merged$phylum) &
                           is.na(rds_mrel_merged$division),"phylum"])
} else {print("ERROR, nonmicstate must be 'remove' or 'add'")}

## REPORT: Save files.
if(isTRUE(savefiles)){
  save(rds_mrel_merged, file = paste0(RSAV, "44_CH4_KEGG_TAXA.rda"))
}

```

# Log-Ratio Proportionality
## NETWORKS

```{r rho_network_gen, include=FALSE, eval=FALSE}
# Run first limma_CLR chunk (Differential Abundance)
toptbl_g.nt <- toptbl$LOWvsHIGH %>% rownames_to_column("Feature") # LOWvsH_MID contrast was added manually

rho_gen <- propr(counts = codata.GBM$genus, metric = "rho", ivar = "clr")
rho_gen_filt <- rho_gen[">=", .4]["<", -.4]
rho_gen_filt <- propr::simplify(rho_gen_filt)

cyt.tbl <- cytescape(rho_gen_filt, prompt = FALSE)
cyt.tbl$corrsize <- abs(cyt.tbl$rho)
cyt.tbl$corrsense <- ifelse(cyt.tbl$rho < 0, "-", "+")

cyt.tbl.feats <- unique(c(cyt.tbl$Partner, cyt.tbl$Pair))
toptbl.attr <- c("Feature", "logFC")
taxtbl.attr <- c("superkingdom", "phylum", "family", "genus")
cyt.attr <- right_join(toptbl_g.nt[,toptbl.attr], distinct(data.tx_tax[,taxtbl.attr]), by = c("Feature" = "genus"))
cyt.attr <- cyt.attr[cyt.attr$Feature %in% cyt.tbl.feats,]
cyt.attr$CH4_assoc <- ifelse(is.na(cyt.attr$logFC), "N/A", 
                             ifelse(cyt.attr$logFC < 0, "LOW", "HIGH"))
cyt.attr[is.na(cyt.attr)] <- ""

cyt.tbl.namedir <- paste0(WD, 'Rsaves/Networks_taxa/Cyt_Edges', '_Corr04', '_gen.tsv')
cyt.attr.namedir <- paste0(WD, 'Rsaves/Networks_taxa/Cyt_Nodes', '_Corr04', '_gen_', qval, '.tsv')

write.table(cyt.tbl, paste0(WD, cyt.tbl.namedir), sep='\t', row.names = F, quote = F)
write.table(cyt.attr, paste0(WD, cyt.attr.namedir), sep='\t', row.names = F, quote = F)

```

```{r rho_network_kegg, include=FALSE, eval=FALSE}
# Run first limma_CLR chunk (Differential Abundance)
# Also add LOWvsH_MID contrast:
toptbl_k.nt <- toptbl_k$LOWvsHIGH %>% rownames_to_column("Feature") %>% mutate(Contrast = "LOWvsHIGH")
toptbl_k.nt <- rbind(toptbl_k.nt, toptbl_k$LOWvsH_MID %>% rownames_to_column("Feature") %>% mutate(Contrast = "LOWvsH_MID"))
toptbl_k.nt <- toptbl_k.nt[!duplicated(toptbl_k.nt$Feature),]

## RHO SUBSETS FOR NETWORKS:
rho_kgg <- propr(counts = codata.GBM_k, metric = "rho", ivar = "clr")
rho_kgg_filt_0.8 <- rho_kgg[">=", .8]["<", -.8]
rho_kgg_filt_0.8 <- propr::simplify(rho_kgg_filt_0.8)
rho_kgg_filt_0.7 <- rho_kgg[">=", .7]["<", -.7]
rho_kgg_filt_0.7 <- propr::simplify(rho_kgg_filt_0.7)

## REPORT: Save files.
if(isTRUE(savefiles)){
  save(toptbl_k.nt, rho_kgg_filt_0.7, rho_kgg_filt_0.8, file = paste0(RSAV, "52_FILT_RHO_KEGG.rda"))
}

# Cytoscape table: edges (correlations)
cyt.tbl <- cytescape(rho_kgg_filt_0.7, prompt = FALSE)
cyt.tbl$corrsize <- abs(cyt.tbl$rho)
cyt.tbl$corrsense <- ifelse(cyt.tbl$rho < 0, "-", "+")

# Cytoscape table: nodes (KEGG)
cyt.tbl.feats <- unique(c(cyt.tbl$Partner, cyt.tbl$Pair))
toptbl.attr <- c("Feature", "logFC", "Contrast")
data.kg_class.nt <- data.kg_class %>% rownames_to_column("KEGG_ID")
classtbl.attr <- c("KEGG_ID", "Association")
cyt.attr <- right_join(toptbl_k.nt[,toptbl.attr], distinct(data.kg_class.nt[,classtbl.attr]), by = c("Feature" = "KEGG_ID"))
cyt.attr <- cyt.attr[cyt.attr$Feature %in% cyt.tbl.feats,]
cyt.attr$CH4_assoc <- ifelse(is.na(cyt.attr$logFC), "N/A", 
                             ifelse(cyt.attr$logFC < 0, "LOW", "HIGH"))
cyt.attr[is.na(cyt.attr)] <- ""

cyt.tbl.namedir <- paste0(WD, 'Rsaves/Networks_kegg/Cyt_Edges', '_Corr07', '_kegg.tsv')
cyt.attr.namedir <- paste0(WD, 'Rsaves/Networks_kegg/Cyt_Nodes', '_Corr07', '_kegg_', qval, '.tsv')

write.table(cyt.tbl, cyt.tbl.namedir, sep='\t', row.names = F, quote = F)
write.table(cyt.attr, cyt.attr.namedir, sep='\t', row.names = F, quote = F)

```

```{r corr_ch4_RA, include=FALSE, eval=FALSE}
# Correlation table from CLR-transformed reads: 
cor_ch4_ra <- data.frame(matrix(ncol = 4, nrow = 0))
colnames(cor_ch4_ra) <- c("X","Y","Corr_est","Pval")
extids <- rownames(ids[which(ids$CH4_Rank %in% c("HIGH", "LOW")),])
y <- setNames(ids$ch4.spkpm_adjusted, rownames(ids))
y <- y[names(y) %in% extids]
xdat <- data.frame(codata.GBM.clr$genus$LR)
xdat <- xdat[extids,]
for(g in 1:ncol(xdat)){
  x <- xdat[,g]
  names(x) <- rownames(xdat)
  if(all.equal(names(x), names(y))){
    corr <- cor.test(x, y, method = "pearson")
    cor_ch4_ra[g,] <- c(colnames(xdat)[g],"ch4.spkpm_adjusted",corr$estimate, corr$p.value)
  }
}
cor_ch4_ra$Padj <- p.adjust(cor_ch4_ra$Pval, method = "BH")
cor_ch4_ra <- cor_ch4_ra[cor_ch4_ra$Pval <= 0.05,]

# Check individual correlations in raw reads dataset:
x <-unlist(data_grps$genus$Counts["Clavispora",extids])
cor.test(x, y, method = "pearson")
plot(x, y)

write.table(cor_ch4_ra, paste0(WD,"corr_ch4_RA_sig.tsv"), sep = "\t", quote = F, row.names = F)

```

## NETWORKS TO CYTOSCAPE

```{r NETWORK_to_Cytoscape_tx, include=FALSE, eval=FALSE}
library(RCy3); library(gplots)
WD <- "D:/Adrian/METALGEN/MinION/"
RSAV <- paste0(WD, "Rsaves/Rdata/")

cyt.tbl <- read.table(paste0(WD, 'Rsaves/Networks_taxa/Cyt_Edges_Corr04_gen.tsv'), sep='\t', stringsAsFactors=F, header=T)
colnames(cyt.tbl)[1:2] <- c("source", "target")
cyt.tbl[,1:2] <- apply(cyt.tbl[,1:2], 2, gsub, pattern = "Unclassified", replacement = "Uncl\\.")
cyt.tbl[,1:2] <- apply(cyt.tbl[,1:2], 2, gsub, pattern = "Candidatus", replacement = "Cdt\\.")
cyt.attr <- read.table(paste0(WD, 'Rsaves/Networks_taxa/Cyt_Nodes_Corr04_gen_00505.tsv'), sep='\t', stringsAsFactors=F, header=T)
colnames(cyt.attr)[1] <- "id"
cyt.attr[,1] <- gsub(pattern = "Unclassified", replacement = "Uncl\\.", cyt.attr[,1])
cyt.attr[,1] <- gsub(pattern = "Candidatus", replacement = "Cdt\\.", cyt.attr[,1])
cyt.attr$Nchar <- nchar(cyt.attr$id)
cyt.attr$LabLength <- ifelse(cyt.attr$Nchar < 7, 50,
                             ifelse(cyt.attr$Nchar %in% c(7:11), 80,
                                    ifelse(cyt.attr$Nchar %in% c(12:16), 100,
                                           ifelse(cyt.attr$Nchar %in% c(17:26), 150, 200))))

createNetworkFromDataFrames(cyt.attr, cyt.tbl, title="Network_Corr04_gen_00505")
#loadTableData(cyt.attr, data.key.column = "id")

setNodeBorderWidthDefault(5)
setNodeBorderColorDefault(gplots::col2hex("black"))
setNodeBorderColorMapping('superkingdom', unique(cyt.attr$superkingdom), sapply(c('purple', 'brown', 'violet'), col2hex), mapping.type = "d")
setNodeShapeDefault('DIAMOND')
setNodeShapeMapping('superkingdom', unique(cyt.attr$superkingdom), c('ROUND_RECTANGLE', 'ELLIPSE', 'RECTANGLE'))
setNodeColorMapping('CH4_assoc',unique(cyt.attr$CH4_assoc), sapply(c('grey', 'orange', 'green'), col2hex), mapping.type = "d")
mapVisualProperty(visual.prop = "Node Z Location", unique(cyt.attr$CH4_assoc), c(0, 1, 1))

bundleEdges()
setEdgeOpacityMapping('corrsize', c(min(cyt.tbl$corrsize), max(cyt.tbl$corrsize)), c(20, 90))
setEdgeColorMapping('corrsense', unique(cyt.tbl$corrsense), sapply(c('blue', 'darkred'), col2hex), mapping.type = "d")

#layoutNetwork('force-directed edgeAttribute=rho defaultSpringLength=70 defaultSpringCoefficient=0.00005')

## ANALYZED NODES OF PRINCIPAL NETWORK:
nds <- read.table(paste0(WD, "Rsaves/Networks_taxa/Network_Corr04_gen_00505 default node.csv"), stringsAsFactors = F, sep = ',', header = T)
nds$DA <- str_replace(nds$CH4_assoc, "LOW|HIGH", "YES")
nds$DA <- str_replace(nds$DA, "N/A", "NO")

t.test(BetweennessCentrality ~ DA, data = nds)

```

```{r NETWORK_to_Cytoscape_kg, include=FALSE, eval=FALSE}
library(RCy3); library(gplots)
WD <- "D:/Adrian/0-INIA/METALGEN/MinION/"
RSAV <- paste0(WD, "Rsaves/Rdata/")

cyt.tbl <- read.table(paste0(WD, 'Rsaves/Networks_kegg/Cyt_Edges_Corr07_kegg.tsv'), sep='\t', stringsAsFactors = F, header = T)
colnames(cyt.tbl)[1:2] <- c("source", "target")
cyt.attr <- read.table(paste0(WD, 'Rsaves/Networks_kegg/Cyt_Nodes_Corr07_kegg_005.tsv'), sep='\t', stringsAsFactors = F, header = T)
colnames(cyt.attr)[1] <- "id"

createNetworkFromDataFrames(cyt.attr, cyt.tbl, title="Network_Corr07_kegg_005")

setNodeBorderWidthDefault(8)
setNodeBorderColorDefault(gplots::col2hex("black"))
setNodeBorderColorMapping('Methane', unique(cyt.attr$Methane), sapply(c('black', 'violet', 'purple'), col2hex), mapping.type = "d")
setNodeShapeDefault('ELLIPSE')
setNodeShapeMapping('Methane', unique(cyt.attr$Methane), c('ELLIPSE', 'RECTANGLE', 'DIAMOND'))
setNodeColorMapping('CH4_assoc', unique(cyt.attr$CH4_assoc), sapply(c('grey', 'green', 'orange'), col2hex), mapping.type = "d")
mapVisualProperty(visual.prop = "Node Z Location", unique(cyt.attr$Methane), c(0, 2, 1))

bundleEdges()
setEdgeOpacityMapping('corrsize', c(min(cyt.tbl$corrsize), max(cyt.tbl$corrsize)), c(20, 90))
setEdgeColorMapping('corrsense', unique(cyt.tbl$corrsense), sapply(c('blue', 'darkred'), col2hex), mapping.type = "d")

#layoutNetwork('force-directed edgeAttribute=rho defaultSpringLength=70 defaultSpringCoefficient=0.00005')

```
